crypto/armcap.c

```diff
@@ -28,6 +28,77 @@ uint32_t OPENSSL_rdtsc(void)
 {
     return 0;
 }
+#elif defined(_WIN32)
+# include <windows.h>
+# include "internal/ctype.h"
+
+static int todigit(WCHAR c)
+{
+    if (ossl_isdigit(c))
+        return c - '0';
+    else if (ossl_isxdigit(c))
+        return ossl_tolower(c) - 'a' + 10;
+
+    /* return largest base value to make caller terminate the loop */
+    return 16;
+}
+
+static unsigned int strtouint(const WCHAR *str)
+{
+    unsigned int digit, base = 10, ret = 0;
+
+    if (*str == '0') {
+        base = 8, str++;
+        if (ossl_tolower(*str) == 'x')
+            base = 16, str++;
+    }
+
+    while((digit = todigit(*str++)) < base)
+        ret = ret * base + digit;
+
+    return ret;
+}
```

Comment:
I've said it before and I'll said it again, we *have* to stop code copying all over the place.  These functions already exist in `crypto/cryptlib.c`, I would much much rather have them become non-static but remain non-exported, possibly both prefixed with `ossl_`.

Reply:
I is probably because of the WCHAR and GetEnvironmentVariableW.
TBH, I am not very happy with those lingering in libcrypto...

Reply:
One should note that both copies appear in platform-specific section, and platforms are not overlapping. So that there is no *binary* code duplication. Yes, there is source code duplication, but what are alternatives? Note that the subroutines are type-variant, in sense that input type is not same, but depends on platform (on its other "dimension"(*)), so that making it static is actually least error-prone way. But is there way to avoid source code duplication? Sure, header. But again, code is type-variant and header would have to depend on external definition, and this is kind of confusing and non-stylish. At least overall it would be less readable. I mean typedef followed by #include is less readable than as it stands now. Because code is small enough to contemplate at once, or in other words fits a screen.

As for having this code at all. Reason for avoiding VCRT and adhering to KERNEL32.DLL interfaces is that the subroutine in question is meant to be used as "constructor" and calling VCRT from "constructor" was reported to cause dead-locks and crashes.

(*) in my mind "platform" is "multidimensional" term, processor being one "dimension" and OS - another.

Reply:
> not seeing any overflow detection in either copy

If one detects it, what would be suggested action? Ignore the value and proceed as if environment variable was not set at all? Or pretend as if it was set to 0? All ones? Rationale behind being lax here is that whomever sets the variable just **has to** know what [s]he is going in either case. Just like you wouln't blame debugger for letting you planting an inappropriate value in a register of your fancy that would trigger crash in next instruction or some time after.

Reply:
> isn't the trick to move its definition to `crypto/include/internal/cryptlib_int.h`

Keyword was "least error-prone."

Reply:
I see no issue with having such functions in `cryptlib.c`...  after all, that's supposed to be a kind of utility library for the rest of libcrypto, right?  A set of `variant_char` routines in there would be *entirely* appropriate.

Reply:
Having another look, I gotta admit that the one that got me reacting the most is `todigit`.  I can understand that `strtouint` isn't exactly like `ossl_strtouint64` from cryptlib.c, considering return type.  Is that where the error-prone discussion comes from?

Reply:
I think the "error-prone" discussion is relating to the fact that, at present, these implementations are tightly coupled to a specific feature; in this file, reading the OPENSSL_armcap environment variable.  It's okay to make shortcuts here (because it's always manually set and if set improperly a footgun), but attempting to generalize and share the function would be error-prone since it loses the tight coupling between implementation and consumer, and the unwritten assumptions could more easily be violated.
That said, I would prefer to write down the "unwritten assumption" to the extent possible, even if the implementation and usage remain tightly coupled for now.

Reply:
Ok, I hear you...